### PR TITLE
[train&val] Support specify validation dataset in WITH clause

### PR DESCRIPTION
**Pull request by weiguoz**
Part job of #1111(NOT Taobao 11.11 Online Shopping Festival).
This PR support specifying validation dataset in `WITH` clause, e.g.
- TensorFlow
``` SQL
SELECT *
FROM iris.train
TRAIN DNNClassifier
WITH
  model.n_classes = 3,
  model.hidden_units = [10, 20],
  validation.select = "SELECT * FROM iris.train limit 20"  -- NOTE: It's a SELECT statement
COLUMN sepal_length, sepal_width, petal_length, petal_width
LABEL class
INTO sqlflow_models.my_dnn_model
```
or 
- ALPS job
```SQL
SELECT deep_id, user_space_stat, user_behavior_stat, space_stat, l
FROM my_train_table
TRAIN DNNClassifier
WITH
    model.n_classes = 2,
    model.hidden_units = [10, 20], 
    engine.ps_num = 0,
    validation.table = "my_validatioin_table" -- NOTE: It's a table name
COLUMN
    SPARSE(deep_id,15033,COMMA,int),
    EMBEDDING(CATEGORY_ID(deep_id,15033,COMMA),512,mean) -- ...
LABEL l
INTO model_table; 
```

### CODE REVIEW

**Comment by terrytangyuan** [comment]
Why does TensorFlow job use a SELECT statement while ALPS use a table name?

**Review by terrytangyuan — COMMENTED** [review]
Why does TensorFlow job use a SELECT statement while ALPS use a table name? What does the final implementation look like? I am concerned that this will lead to a lot of inconsistencies among different engines.

**Comment by weiguoz** [comment]
> Why does TensorFlow job use a SELECT statement while ALPS use a table name? What does the final implementation look like? I am concerned that this will lead to a lot of inconsistencies among different engines.
1. For the first question
> Why does TensorFlow job use a SELECT statement while ALPS use a table name?

The `SELECT` statement provides a stronger expressiveness. It should have been the only way.
But the `alps.io.reader.odps_reader.OdpsReader` does not support a `SELECT` statement but a table name.
https://github.com/sql-machine-learning/sqlflow/blob/64b6219c9d927c88db54677fa1b4e1006b28218c/pkg/sql/template_alps.go#L96-L103
I'm asking the alps team if there could support the SELECT statement instead of a table name.

2. For the 2nd one
> What does the final implementation look like?

If we can't unify to the `SELECT` statement for some reason at last. I would propose the below 2 methods as discussed in issue #1111
```SQL
SELECT ... TO TRAIN DNNClassifier ... INTO my_model
VALIDATE ON (SELECT * FROM my_validation_table WHERE random>0.8) -- By SELECT statement
```
OR
```SQL
SELECT ... TO TRAIN DNNClassifier ... INTO my_model
VALIDATE ON TABLE my_validation_table -- By Table name
```

**Comment by shendiaomo** [comment]
> Part job of #1111(NOT Taobao 11.11 Online Shopping Festival).
> This PR support specifying validation dataset in `WITH` clause, e.g.
> 
> * TensorFlow
> 
> ```sql
> SELECT *
> FROM iris.train
> TRAIN DNNClassifier
> WITH
>   model.n_classes = 3,
>   model.hidden_units = [10, 20],
>   validation.dataset = "SELECT * FROM iris.train limit 20"  -- NOTE: It's a SELECT statement
> COLUMN sepal_length, sepal_width, petal_length, petal_width
> LABEL class
> INTO sqlflow_models.my_dnn_model
> ```
> 
> or
> 
> * ALPS job
> 
> ```sql
> SELECT deep_id, user_space_stat, user_behavior_stat, space_stat, l
> FROM my_train_table
> TRAIN DNNClassifier
> WITH
>     model.n_classes = 2,
>     model.hidden_units = [10, 20], 
>     engine.ps_num = 0,
>     validation.dataset.table = "my_validatioin_table" -- NOTE: It's a table name
> COLUMN
>     SPARSE(deep_id,15033,COMMA,int),
>     EMBEDDING(CATEGORY_ID(deep_id,15033,COMMA),512,mean) -- ...
> LABEL l
> INTO model_table; 
> ```

How about supporting both table and select statement for all the job types? Tables and select statements are mostly interchangeable in standard SQL. Mimicking this mechanism will make the learning curve less steep for standard SQL users. For jobs other than alps, a `table` can be substituted with a `select * from table`statement, and for alps jobs, a `select` statement can be substituted with a temporary table that is passed to alps configuration.

File: cmd/sqlflowserver/main_test.go [hunk 1]
```diff
@@ -821,21 +826,31 @@ INTO trained_elasticdl_keras_classifier;`, os.Getenv("MAXCOMPUTE_PROJECT"), "sql
...
+	LIMIT 100
+	TRAIN DNNClassifier
+	WITH
+	    model.n_classes = 2,
+	    model.hidden_units = [10, 20],
+	    train.batch_size = 10,
+	    engine.ps_num = 0,
+	    engine.worker_num = 0,
+	    engine.type = local,
+	    validation.dataset.table = "%s.sparse_column_test"
```
Comment (tonyyang-svail) [inline cmd/sqlflowserver/main_test.go]: Naming: `validation.dataset.table` or `validation.table`?

File: pkg/sql/create_train_val.go [hunk 2]
```diff
@@ -1,200 +0,0 @@
-// Copyright 2019 The SQLFlow Authors. All rights reserved.
```
Comment (tonyyang-svail) [inline pkg/sql/create_train_val.go:1]: ~~Is it possible to keep this feature? A user may enable it by `validation.split_ratio = 0.2`.~~

This feature requires SQLFlow to modify the database by creating temporary tables during the execution of the Go code, while the Go code should only be responsible for parsing and code generation. The execution is left for the workflow engine.

File: pkg/sql/ir_generator.go [hunk 3]
```diff
@@ -567,6 +568,28 @@ func parseShape(e *expr) ([]int, error) {
...
+			if len(subk[1]) > 0 {
+				g[subk[1]] = v
+			}
+		}
+	}
+	return g
+}
+
+func parseValidataionDataset(attrs map[string]interface{}) (string, error) {
+	validation := parseAttrsGroup(attrs, "validation.")
```
Comment (tonyyang-svail) [inline pkg/sql/ir_generator.go:585]: I am not sure if we need to check the existence of the "dataset" key.

File: pkg/sql/ir_generator.go [hunk 4]
```diff
@@ -59,12 +59,13 @@ func generateTrainIR(slct *extendedSelect, connStr string) (*codegen.TrainIR, er
 			Name: tc.label,
 		}}
 
-	// TODO(typhoonzero): fill in ValidationSelect using `create_train_val.go`
-	// TODO(typhoonzero): fill in ValidationSelect when VALIDATE clause is ready
+	validationDS, _ := parseValidataionDataset(attrList)
```
Comment (tonyyang-svail) [inline pkg/sql/ir_generator.go]: Naming: Does `DS` refer to `Dataset`? If so, maybe change to `validationDataset` or `validationSelect`.

File: cmd/sqlflowserver/main_test.go [hunk 5]
```diff
@@ -573,13 +573,18 @@ func CaseSelect(t *testing.T) {
...
-LABEL class
-INTO sqlflow_models.my_dnn_model;`, caseDB, caseTrainTable)
+	trainSQL := fmt.Sprintf(`
+	SELECT *
+	FROM %s.%s
+	TRAIN DNNClassifier
+	WITH
+		model.n_classes = 3,
+		model.hidden_units = [10, 20],
+		validation.dataset = "SELECT * FROM %s.%s LIMIT 30"
```
Comment (tonyyang-svail) [inline cmd/sqlflowserver/main_test.go]: Naming:

- `validation.dataset` or `validation.select`?

- `validate` or `validation`?

    - verb: train, predict, validate

    - noun: training, prediction/predicting, validation

**Review by tonyyang-svail — DISMISSED** [review]
@weiguoz Thanks for this PR. LGTM.

I have left some comments on possible improvements.

**Comment by tonyyang-svail** [comment]
> Why does TensorFlow job use a SELECT statement while ALPS use a table name? What does the final implementation look like? I am concerned that this will lead to a lot of inconsistencies among different engines.

Hi @terrytangyuan, thanks for mentioning this. This inconsistency has been concerning since the first day we worked on supporting ALPS on ODPS.

As @weiguoz has mentioned. If we have to support only one of `select` or `table`. We will go for the `select`. And the codegen of ALPS/EDL will add a step of creating a temporary table based on the `select` in the generated code.

File: cmd/sqlflowserver/main_test.go (hunk 1 above)
Comment (typhoonzero) [reply cmd/sqlflowserver/main_test.go]: +1 for `validation.table`

**Comment by weiguoz** [comment]
Specify the dataset for validation by:
1. select
```SQL
validation.select= "SELECT * FROM my_validation_table limit 20"
```
2. table name
```SQL
validation.table="my_validation_table"
```

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**